File: Demo/Swift/语法/16.Swift3.0 - 循环引用/16.Swift3.0 - 循环引用/ViewController.m
//
//  ViewController.m
//  16.Swift3.0 - 循环引用
//
//  Created by Hayder on 16/10/14.
//  Copyright © 2016年 limpid. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (nonatomic, copy) void (^cpmpleCallBack)() ;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //解除引用1: __weak
//    __weak typeof (self) weakSelf = self;
//    [self loadData:^{
//       
//        NSLog(@"%@",weakSelf.view);
//    }];
    
    //解除引用2: __unsafe_unretained
    //高级iOS 程序员如果需要自行管理内存，可以考虑使用，但是不建议使用
    __unsafe_unretained typeof (self) weakSelf = self;
    
    //EXC_BAD_ACCESS 坏内存访问，野指针访问
    //__unsafe_unretained 同样是 assign 的引用(MRC中没有weak)
    //在MRC中如果要弱引用对象，都是使用assign,不会使用引用计数，但是一旦对象被释放，地址不会改变，继续访问，出现野指针
    //在ARC weak,本质上是一个观察者模式，一旦发现对象被释放，会自动将地址设置为nil,更加安全
    //效率: weak 的效率会略微差一些!
    [self loadData:^{
        
        NSLog(@"%@",weakSelf.view);
    }];
}

- (void)loadData:(void (^)())completion
{
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
       
        //使用属性记录block
        self.cpmpleCallBack = completion;
        NSLog(@"耗时操作");
        
        [NSThread sleepForTimeInterval:2.0];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            
            //执行block
            completion();
        });
        
    });
}

@end

File: Demo/Swift/语法/14.Swift3.0 - 闭包/14.Swift3.0 - 闭包/ViewController.swift
//
//  ViewController.swift
//  14.Swift3.0 - 闭包
//
//  Created by WZH on 16/10/14.
//  Copyright © 2016年 limpid. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    //类似于OC的block，但比OC的block应用更加的广泛
    //区别: 在OC中block是匿名的函数
    //      在Swift中函数是特殊的闭包
    
    /**
        闭包
        1.提前准备好的代码，在需要的时候执行
        2.可以当做参数传递
     */
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //1> 闭包
        //() -> () 没有参数，没有返回值的函数
        //如果没有参数，没有返回值，可以省略，连in一起省略
        let b1 = {
            
            print("hello")
        }
        
        //执行闭包
        b1()
        
        //2. 带参数的闭包
        //(Int) -> ()
        //闭包中，参数，返回值，实现代码都是写在{}中
        //需要使用一个关键字 “in” 分隔定义和实现
        //{ (形参列表 -> 返回值类型 in 实现代码)}
        let b2 = {
            
            (x: Int) -> ()
            
            in
            
            print(x)
        }
        
        b2(100)
        
        //3.带参数/返回值的闭包
        //类型:(Int) -> Int
        let b3 = { (x: Int) -> Int in
            
            return x
        }
        
        print(b3(10))
    }

    func demo() {
        
        print(sum(x:10, y:20))
        
        //1.定义一个常量来记录函数
        let f = sum
        
        //2.在需要的时候执行
        print(f(10, 20))
    }
    func sum(x: Int, y: Int) ->Int{
        
        return x + y
    }
}

